### PR TITLE
Production update

### PR DESCRIPTION
**Pull request by JuusoMRusanen**

### CODE REVIEW

File: src/utils/toolbar-utils.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        setToolbarFormMode(ToolbarFormModes.CREATE);
+        break;
+      case selectedRowIds.length === 1:
+        setToolbarFormMode(ToolbarFormModes.EDIT);
+        break;
+      default:
+        setToolbarFormMode(ToolbarFormModes.NONE);
+    }
+  }
+};
```
Comment (JuusoMRusanen) [inline src/utils/toolbar-utils.ts:36]: Could do without an interface here.

**Review by JuusoMRusanen — COMMENTED** [review]
Some improvements could be made

File: src/components/home/balance-card.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+const BalanceCard = () => {
+  const persons = useAtomValue(personsAtom);
+  const userProfile = useAtomValue(userProfileAtom);
+  const [timespan, setTimespan] = useAtom(timespanAtom);
+  const { personsApi } = useApi();
+  const setError = useSetAtom(errorAtom);
+  const [loading, setLoading] = useState(false);
+  const [personTotalTime, setPersonTotalTime] = useAtom(personTotalTimeAtom);
+  const adminMode = UserRoleUtils.adminMode();
+  const beforeDate = DateTime.now().minus({ days: 1 });
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/home/balance-card.tsx:30]: renamed

File: src/components/timebank/timebank-content.tsx [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            <Box
+              sx={{
+                display: "flex",
+                flexDirection: "row",
+                justifyItems: "center"
+              }}
+            >
+              {renderDailyEntryOrRangeChart()}
+              <List dense sx={{ marginLeft: "5%" }}>
+                <ListItem>
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:308]: Solved

File: src/components/vacation-requests-table/vacation-requests-table-columns.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import { DateTime } from "luxon";
+import { useAtomValue } from "jotai";
+import { languageAtom } from "../../atoms/language";
+import LocalizationUtils from "../../utils/localization-utils";
+
+/**
+ * Vacation requests table columns component
+ */
+const VacationRequestsTableColumns = () => {
+  const language = useAtomValue(languageAtom);
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-columns.ts:12]: Removed const language and formatDate as it already exists in the utils

File: src/components/vacation-requests-table/vacation-requests-table-columns.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+   * @param date datetime object
+   * @param dateWithTime datetime object with time
+   * @returns formatted date time
+   */
+  const formatDate = (date: DateTime, dateWithTime?: boolean) => {
+    if (!date) return "";
+    return date
+      .setLocale(language)
+      .toLocaleString(dateWithTime ? DateTime.DATETIME_SHORT : undefined);
+  };
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-columns.ts:26]: Removed and imported instead

File: src/components/vacation-requests-table/vacation-requests-table.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  const vacationRequests = useAtomValue(adminMode ? allVacationRequestsAtom : vacationRequestsAtom);
+  const vacationRequestStatuses = useAtomValue(
+    adminMode ? allVacationRequestStatusesAtom : vacationRequestStatusesAtom
+  );
+  const containerRef = useRef(null);
+  const [formOpen, setFormOpen] = useState(false);
+  const [selectedRowIds, setSelectedRowIds] = useState<GridRowSelectionModel>([]);
+  const createDataGridRows = VacationRequestsTableRows();
+  const [rows, setRows] = useState<VacationsDataGridRow[]>([]);
+  const language = useAtomValue(languageAtom);
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table.tsx:62]: This is actually needed here for the useMemo hook. Leaving it.

File: src/utils/chart-utils.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    nonBillableProject: personTotalTime.nonBillableProjectTime
+  },
+  { name: strings.timebank.expected, expected: personTotalTime.expected }
+];
+
+/**
+ * Renders custom labels in the pie chart
+ *
+ * @param props Props from the pie chart data, such as name and value
+ * @returns string
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/utils/chart-utils.tsx:42]: changed to @returns custom label string

File: src/utils/toolbar-utils.ts (hunk 1 above)
Comment (JuusoMRusanen) [reply src/utils/toolbar-utils.ts:36]: Re-implemented without an interface

File: package.json [hunk 8]
```diff
@@ -17,25 +17,32 @@
     "@mui/icons-material": "^5.14.7",
     "@mui/material": "^5.14.7",
     "@mui/types": "^7.2.9",
+    "@mui/x-data-grid": "^6.13.0",
+    "@mui/x-date-pickers": "^6.14.0",
     "@types/luxon": "^3.3.2",
     "@types/react-router": "^5.1.20",
     "@types/react-router-dom": "^5.3.3",
+    "@types/recharts": "^1.8.24",
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply package.json:25]: Moved type definitions to devDependencies

File: src/components/charts/timebank-multibarchart.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  selectedEntries: DailyEntry[];
+}
+
+/**
+ * Timebank multiple bar chart component representing data from a selected range
+ *
+ * @param props component properties
+ */
+const TimebankMultiBarChart = ({selectedEntries}: Props) => {
+  const chartData = selectedEntries.map((entry) => ({
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/charts/timebank-multibarchart.tsx:23]: Well, now it is.

File: src/components/timebank/timebank-content.tsx [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+                    width: "100%"
+                  }}
+                  value={timespan}
+                  onChange={(e) => {
+                    setTimespan(e.target.value as Timespan);
+                  }}
+                >
+                  <MenuItem value={Timespan.WEEK}>{strings.timeExpressions.week}</MenuItem>
+                  <MenuItem value={Timespan.MONTH}>{strings.timeExpressions.month}</MenuItem>
+                  <MenuItem value={Timespan.ALL_TIME}>{strings.timeExpressions.allTime}</MenuItem>
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:229]: It's now mapped like suggested and includes yearly timespan

File: src/components/timebank/timebank-content.tsx [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+              <FormControl
+                sx={{
+                  width: "50%",
+                  marginRight: "1%",
+                  textAlign: "center"
+                }}
+              >
+                <InputLabel id="select">{strings.timebank.selectTimespan}</InputLabel>
+                <Select
+                  label="Select a time span"
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:217]: Localized

File: src/components/charts/timebank-overviewchart.tsx [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import { personTotalTimeAtom } from "../../atoms/person";
+import { useAtomValue } from "jotai";
+import { Worktime } from "../../types";
+
+/**
+ * Time bank overview chart component
+ */
+const TimebankOverviewChart = () => {
+  // rome-ignore lint/style/noNonNullAssertion: personTotalTime is validated in the parent component with conditional return
+  const personTotalTime = useAtomValue(personTotalTimeAtom)!;
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/charts/timebank-overviewchart.tsx:15]: Made it a prop like discussed

File: src/components/charts/timebank-piechart.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import { personDailyEntryAtom } from "../../atoms/person";
+import { useAtomValue } from "jotai";
+import { COLORS } from "../constants";
+
+/**
+ * Time bank pie chart component
+ * @returns A pie chart containing logged time breakdown of the daily entry
+ */
+const TimebankPieChart = () => {
+  const personDailyEntry = useAtomValue(personDailyEntryAtom);
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/charts/timebank-piechart.tsx:18]: Made it a prop like discussed

File: src/components/contexts/confirmation-handler.tsx [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      onCancel={() => setOpen(false)}
+      onConfirm={() => {
+        setOpen(false);
+        deleteVacationsData();
+      }}
+      confirmButtonText={strings.confirmationHandler.confirmButtonText}
+      cancelButtonText={strings.confirmationHandler.cancelButtonText}
+      title={strings.confirmationHandler.title}
+    >
+      {open && (
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/contexts/confirmation-handler.tsx:34]: Removed redundant conditional

File: src/components/contexts/error-handler.tsx [hunk 15]
```diff
@@ -68,14 +68,11 @@ const ErrorHandler = ({ children }: Props) => {
...
       >
-        <DialogContent id="error-dialog-description">
-          {error && (
-            <Typography marginBottom={3} sx={{ fontSize: 16, fontWeight: "bold" }}>
-              {error}
-            </Typography>
-          )}
-          <code style={{ fontSize: "12px" }}>{error || ""}</code>
-        </DialogContent>
+        {error && (
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/contexts/error-handler.tsx:71]: Removed redundant conditional

File: src/components/generics/date-range-picker.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+   * Handle date change and calculate days
+   *
+   * @param props startDate, endDate, days
+   */
+  const handleDateChange = (startDate: DateTime, endDate: DateTime) => {
+    const updatedEndDate = startDate > endDate ? startDate : endDate;
+    setStartDate(startDate);
+    setEndDate(updatedEndDate);
+    setDates(startDate, updatedEndDate, getVacationDurationInDays(startDate, updatedEndDate));
+  };
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/generics/date-range-picker.tsx:41]: Placed a useMemo in the parent component to set vacationData and got rid of the setDates function all together

File: src/components/home/balance-card.tsx [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    <Link to={"/timebank"} style={{ textDecoration: "none" }}>
+      <Card
+        sx={{
+          "&:hover": {
+            background: "#efefef"
+          }
+        }}
+      >
+        <CardContent>
+          <h3 style={{ marginTop: 6 }}>{strings.timebank.balance}</h3>
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/home/balance-card.tsx:102]: Using Typography now

File: src/components/timebank/timebank-content.tsx [hunk 18]
```diff
@@ -0,0 +1,381 @@
...
+  loading: boolean;
+}
+
+/**
+ * Component that contains the entirety of Timebank content, such as charts
+ *
+ * @param props Component properties
+ */
+const TimebankContent = (props: Props) => {
+  const { handleDailyEntryChange, loading } = props;
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:51]: Spreaded the props

File: src/components/timebank/timebank-content.tsx [hunk 19]
```diff
@@ -0,0 +1,381 @@
...
+  const [selectedEntries, setSelectedEntries] = useState<DailyEntry[]>([]);
+  const [byRange, setByRange] = useState({
+    dailyEntries: false
+  });
+  const personTotalTime = useAtomValue(personTotalTimeAtom);
+  const [timespan, setTimespan] = useAtom(timespanAtom);
+  const personDailyEntry = useAtomValue(personDailyEntryAtom);
+  const dailyEntries = useAtomValue(dailyEntriesAtom);
+  const todayOrEarlier = DateTime.fromJSDate(
+    dailyEntries.filter((item) => item.date <= new Date() && item.logged)[0].date
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:62]: Added a check for length

[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:62]: Made a conditional checking for the dailyEntries length

File: src/components/timebank/timebank-content.tsx [hunk 20]
```diff
@@ -0,0 +1,381 @@
...
+  );
+
+  /**
+   * Allows only logged dates or with expected hours to be selected in the date time picker.
+   *
+   * @param date DateTime object passed from the date picker
+   */
+  const disableNullEntries = (date: DateTime) => {
+    const loggedDates = dailyEntries.find(
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:71]: Removed the function. I think it's appropriate to let the user check any time entry, whether it contains anything or not.

File: src/components/timebank/timebank-content.tsx [hunk 21]
```diff
@@ -0,0 +1,381 @@
...
+        today={todayOrEarlier}
+      />
+    );
+  };
+
+  return (
+    <>
+      <Grow in>
+        <Card sx={{ border: "2px solid #bdbdbd;" }}>
+          <Typography
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:180]: Made a styled component

File: src/components/timebank/timebank-content.tsx [hunk 22]
```diff
@@ -0,0 +1,381 @@
...
+            <Box sx={{ display: "flex", justifyContent: "center" }}>
+              <FormControl
+                sx={{
+                  width: "50%",
+                  marginRight: "1%",
+                  textAlign: "center"
+                }}
+              >
+                <InputLabel id="select">{strings.timebank.selectTimespan}</InputLabel>
+                <Select
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/timebank/timebank-content.tsx:216]: Changed to TextField

File: src/components/vacation-requests-table/skeleton-table-rows/skeleton-table-row-item.tsx [hunk 23]
```diff
@@ -0,0 +1,32 @@
+import { Skeleton } from "@mui/material";
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/skeleton-table-rows/skeleton-table-row-item.tsx:1]: Renamed

File: src/components/vacation-requests-table/vacation-requests-table-rows.ts [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+} from "../../generated/client";
+import LocalizationUtils from "../../utils/localization-utils";
+import strings from "../../localization/strings";
+import { useAtomValue } from "jotai";
+import { personsAtom } from "../../atoms/person";
+import { userProfileAtom } from "../../atoms/auth";
+import { getVacationRequestPersonFullName } from "../../utils/vacation-request-utils";
+
+/**
+ * Vacation requests table rows component
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-rows.ts:16]: Brought the function over to the parent component and removed this file

File: src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+      startDate: startDate,
+      endDate: endDate,
+      days: days
+    });
+  };
+
+  return (
+    <FormControl sx={{ width: "100%" }}>
+      <FormLabel>{strings.vacationRequest.type}</FormLabel>
+      <Select
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx:68]: Changed to TextField

File: src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+          if (vacationType) {
+            setVacationData({
+              ...vacationData,
+              type: vacationType
+            });
+          }
+        }}
+        sx={{ marginBottom: "5px", width: "100%" }}
+      >
+        {(Object.keys(VacationType) as Array<keyof typeof VacationType>).map((vacationType) => {
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx:82]: Removed the casting of the array, however, the vacationType still needs to be casted as VacationType below.

[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx:82]: Removed the casting of the array, however, the vacationType still needs to be casted as VacationType below.

File: src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+            </MenuItem>
+          );
+        })}
+      </Select>
+      <FormLabel>{strings.vacationRequest.message}</FormLabel>
+      <TextField
+        required
+        error={!vacationData.message?.length}
+        value={vacationData.message}
+        onChange={(event: ChangeEvent<HTMLInputElement>) => {
```
[the remark this answers is not in the document]
Comment (JuusoMRusanen) [reply src/components/vacation-requests-table/vacation-requests-table-toolbar/toolbar-form/toolbar-form-fields.tsx:95]: Moved into it's own function

**Review by villejuutila — APPROVED** [review]

**Pull request merged**